galley/cmd/gals/cmd/validator.go

```diff
@@ -0,0 +1,149 @@
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/util"
+)
+
+// waitSignal awaits for SIGINT or SIGTERM and closes the channel
+func waitSignal(stop chan struct{}) {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	<-sigs
+	close(stop)
+	_ = log.Sync()
+}
+
+func validatorCmd(printf, fatalf shared.FormatFn) *cobra.Command {
+	var (
+		webhookConfigName   string
+		pilotWebhookName    string
+		mixerWebhookName    string
+		port                uint
+		domainSuffix        string
+		certFile            string
+		keyFile             string
+		caFile              string
+		healthCheckInterval time.Duration
+		healthCheckFile     string
+	)
+
+	patchCert := func(stop chan struct{}) error {
+		caCertPem, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return err
+		}
+
+		timeout := time.After(time.Minute)
+		cl := common.client.AdmissionregistrationV1beta1().ValidatingWebhookConfigurations()
+		for {
+			webhooks := []string{pilotWebhookName, mixerWebhookName}
+			if err = util.PatchValidatingWebhookConfig(cl, webhookConfigName, webhooks, caCertPem); err == nil {
+				return nil
+			}
+
+			log.Errorf("Patching caBundle in ValidatingWebhookConfigurations %v failed: %v -- retrying",
+				webhookConfigName, err)
+			select {
+			case <-stop:
+				// canceled by caller
+				return nil
+			case <-time.After(time.Second):
+				// retry
+			case <-timeout:
+				return fmt.Errorf("timed out trying to patch ValidatingWebhookConfiguration %q {%q and %q})",
+					webhookConfigName, pilotWebhookName, mixerWebhookName)
+			}
+		}
+	}
+
+	validatorCmd := &cobra.Command{
+		Use:   "validator",
+		Short: "Runs an https server for Istio configuration validation.",
+		Long: "Runs an https server for Istio configuration validation. " +
+			"Uses k8s validating webhooks to validate Pilot and Mixer configuration.",
+		Run: func(cmd *cobra.Command, args []string) {
+			mixerValidator, err := validation.CreateMixerValidator(flags.kubeConfig)
+			if err != nil {
+				fatalf("cannot create mixer backend validator for %q: %v", flags.kubeConfig, err)
+			}
+
+			params := validation.WebhookParameters{
+				MixerValidator:      mixerValidator,
+				PilotDescriptor:     bootstrap.ConfigDescriptor,
+				DomainSuffix:        domainSuffix,
+				Port:                port,
+				CertFile:            certFile,
+				KeyFile:             keyFile,
+				HealthCheckInterval: healthCheckInterval,
+				HealthCheckFile:     healthCheckFile,
+			}
+			wh, err := validation.NewWebhook(params)
+			if err != nil {
+				fatalf("cannot create validation webhook service: %v", err)
+			}
+
+			// Create the stop channel for all of the servers.
+			stop := make(chan struct{})
+
+			go func() {
+				if err := patchCert(stop); err != nil {
+					log.Errorf("Failed to patch webhook config: %v", err)
+				}
+			}()
+
+			go wh.Run(stop)
+			waitSignal(stop)
+		},
+	}
+
+	validatorCmd.PersistentFlags().StringVar(&webhookConfigName,
```

Comment:
Why do we need the webhook-name flags? Seems like these should be constant.

Reply:
For testing purposes in a shared cluster we might need to scope Istio to a single namespace. Multiple validation webhooks would exist and their names ought to be unique.